channel_test.go

```diff
@@ -0,0 +1,186 @@
+package gorums
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/relab/gorums/ordering"
+	"github.com/relab/gorums/tests/mock"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/encoding"
+)
+
+type mockSrv struct{}
+
+func (mockSrv) Test(ctx ServerCtx, _ *mock.Request) (resp *mock.Response, err error) {
+	return nil, nil
+}
+
+func dummyMgr() *RawManager {
+	return NewRawManager(
+		WithGrpcDialOptions(
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+		),
+	)
+}
+
+var handlerName = "mock.Server.Test"
+
+func dummySrv() *Server {
+	mockSrv := &mockSrv{}
+	srv := NewServer()
+	srv.RegisterHandler(handlerName, func(ctx ServerCtx, in *Message, finished chan<- *Message) {
+		req := in.Message.(*mock.Request)
+		defer ctx.Release()
+		resp, err := mockSrv.Test(ctx, req)
+		SendMessage(ctx, finished, WrapMessage(in.Metadata, resp, err))
+	})
+	return srv
+}
+
+func TestChannelCreation(t *testing.T) {
+	node, err := NewRawNode("127.0.0.1:5000")
+	if err != nil {
+		t.Fatal(err)
+	}
+	mgr := dummyMgr()
+	defer mgr.Close()
+	// a proper connection should NOT be established here
+	node.connect(mgr)
+
+	replyChan := make(chan response, 1)
+	go func() {
+		md := &ordering.Metadata{MessageID: 1, Method: handlerName}
+		req := request{ctx: context.Background(), msg: &Message{Metadata: md, Message: &mock.Request{}}}
+		node.channel.enqueue(req, replyChan, false)
+	}()
+	select {
+	case <-replyChan:
+	case <-time.After(3 * time.Second):
+		t.Fatal("deadlock: impossible to enqueue messages to the node")
+	}
+}
+
+func TestChannelSuccessfulConnection(t *testing.T) {
+	addrs, teardown := TestSetup(t, 1, func(_ int) ServerIface {
+		return dummySrv()
+	})
+	defer teardown()
+	mgr := dummyMgr()
+	defer mgr.Close()
+	node, err := NewRawNode(addrs[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err = mgr.AddNode(node); err != nil {
+		t.Fatal(err)
+	}
+	if len(mgr.Nodes()) < 1 {
+		t.Fatal("the node was not added to the configuration")
+	}
+	if !node.channel.isConnected() {
+		t.Fatal("a connection could not be made to a live node")
+	}
+	if node.conn == nil {
+		t.Fatal("connection should not be nil")
+	}
+}
+
+func TestChannelUnsuccessfulConnection(t *testing.T) {
+	mgr := dummyMgr()
+	defer mgr.Close()
+	// no servers are listening on the given address
+	node, err := NewRawNode("127.0.0.1:5000")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// the node should still be added to the configuration
+	if err = mgr.AddNode(node); err != nil {
+		t.Fatal(err)
+	}
+	if len(mgr.Nodes()) < 1 {
+		t.Fatal("the node was not added to the configuration")
+	}
+	if node.conn == nil {
+		t.Fatal("connection should not be nil when NOT using WithBlock()")
+	}
+}
+
+func TestChannelReconnection(t *testing.T) {
+	if encoding.GetCodec(ContentSubtype) == nil {
```

Comment:
I tried without this, and it didn't seem to impact things (tests passed), but I'm unsure why. It will be included in `_gorums.pb.go` files, so if you use `NewManager` from such a generated file, it will be initialized with the `init()` function. Maybe you could tweak and reuse the `tests/dummy/dummy.proto` file instead of adding another proto file. See also the `mgr_test.go` file. Or see my comment on the `mock.proto` file.